src/ripple/overlay/impl/ProtocolMessage.h

```diff
@@ -63,33 +61,143 @@ protocolMessageName (int type)
     default:
         break;
-    };
+    }
     return "unknown";
 }
 
 namespace detail {
 
-template <class T, class Buffers, class Handler>
-std::enable_if_t<std::is_base_of<
-    ::google::protobuf::Message, T>::value,
-        boost::system::error_code>
-invoke (int type, Buffers const& buffers,
+struct MessageHeader
+{
+    /** The size of the message on the wire.
+
+        @note This is the sum of sizes of the header and the payload.
+    */
+    std::uint32_t total_wire_size = 0;
+
+    /** The size of the header associated with this message. */
+    std::uint32_t header_size = 0;
+
+    /** The size of the payload on the wire.
+
+        @note If the message is compressed, this value represents the
+              compressed size.
+    */
+    std::uint32_t payload_wire_size = 0;
+
+    /** The size of the payload.
+
+        @note If compression or other message packing techniques are
+              used, this value represents the size of they payload
+              after decompression.
+    */
+    std::uint32_t payload_real_size = 0;
+
+    /** The type of the message. */
+    std::uint16_t message_type = 0;
+
+    /** Indicates whether the message is compressed. */
+    bool compressed = false;
+
+    /** Indicates which compression algorithm the payload is compressed with. */
+    int algorithm = 0;
+};
+
+template <class BufferSequence>
+boost::optional<MessageHeader> parseMessageHeader(
+    BufferSequence const& bufs,
+    std::size_t size)
+{
+    auto iter = boost::asio::buffers_iterator<BufferSequence, std::uint8_t>::begin(bufs);
+
+    MessageHeader hdr;
+
+    // Version 1 header: uncompressed payload.
+    // The top six bits of the first byte are 0.
+    if ((*iter & 0xFC) == 0)
+    {
+        hdr.header_size = 6;
+
+        if (size < hdr.header_size)
+            return {};
+
+        hdr.compressed = false;
+        hdr.algorithm = 0;
+
+        for (int i = 0; i != 4; ++i)
+            hdr.payload_wire_size = (hdr.payload_wire_size << 8) + *iter++;
+
+        hdr.payload_real_size = hdr.payload_wire_size;
+
+        hdr.total_wire_size = hdr.header_size + hdr.payload_wire_size;
+
+        for (int i = 0; i != 2; ++i)
+            hdr.message_type = (hdr.message_type << 8) + *iter++;
+
+        return hdr;
+    }
+
+    // Version 2 header: compressed payload.
+    if ((*iter & 0x90) == 0x80)
+    {
+        hdr.header_size = 10;
+
+        if (size < hdr.header_size)
+            return {};
+
+        hdr.compressed = true;
+        hdr.algorithm = (*iter & 0x60) >> 5;
+
+        for (int i = 0; i != 4; i++)
+            hdr.payload_wire_size = (hdr.payload_wire_size << 8) + *iter++;
+
+        hdr.payload_wire_size &= 0x0FFFFFFF;
+
+        for (int i = 0; i != 4; i++)
+            hdr.payload_real_size = (hdr.payload_real_size << 8) + *iter++;
+
+        hdr.payload_real_size &= 0x0FFFFFFF;
+
+        hdr.total_wire_size = hdr.header_size + hdr.payload_wire_size;
+
+        for (int i = 0; i != 2; ++i)
```

Comment:
I've gotten used to the `++variable` form unless I _specifically_ need the pre-increment value. I doubt it really matters in practice; compilers are [SMRT](https://www.youtube.com/watch?v=9scdimuHGh4)...

Reply:
From my experience ++i is usually preferred over i++ unless needed anyways. A while back there was discussion pertaining to a rippled code style guide, did that ever materialize?

Reply:
I'm just wondering since in all 3 loops `i` isn't used, but 2 use post-increment and this one pre-increment. Seemed odd, but is apparently more personal style.

Reply:
Huh, I didn't notice that some used pre and others did post. I agree that's a weird. I'll pick one and stick with it.